[BAPL-1893] Migrate to Quarkus (#2)
kiegroup/process-migration-service: README.md

## Requirements

> * JRE 1.8 or 11+

Reviewer: I think it should only be JRE 11+ only because of this:

https://github.com/kiegroup/process-migration-service/blob/26ccb1d245e81a5c4a9d9d194ef666012a6ad1cb/pom.xml#L14-L15

And because Quarkus 2.x is jdk 11 by defaut

Reviewer 2: +1 Quarkus does not support anything older than 11

* Running KIE Process server

### Requirements for testing

PIM Integration Tests requires an external Kie Server instance, for that you can provide your own instance

> or use TestContainers. TestContainers on Podman requires to connect through the external API so make sure

Reviewer: AFAIK Testcontainers on Podman are not officially supported

Author: It is just a comment in case a developer wants to use Podman instead of Docker

to see an example. The provided configuration will be added or override the one existing in application.yaml

## Configuration

Default configuration is as follows:

```yaml
quarkus:
  http:
    auth:
      basic: true
      policy:
        main-policy:
          roles-allowed: admin
      permission:
        main:
          paths: /*
          policy: main-policy
        public:
          paths: /q/health/*
          policy: permit
          methods: GET
  security:
    users:
      file:
        realm-name: pim_file
        enabled: true
        plain-text: true
        users: users.properties
        roles: roles.properties
    jdbc:
      realm-name: Quarkus_jdbc
      enabled: true
      principal-query:
        sql: SELECT u.password, u.role FROM users u WHERE u.username=?
    ldap:
      realm-name: pim_ldap
      dir-context:
        url: ldap://override-when-needed
      identity-mapping:
        search-base-dn: ou=users,o=YourCompany,c=ES
# Quartz configuration
  quartz:
    store-type: jdbc-cmt
    start-mode: forced
  # flyway to create Quartz tables
  flyway:
    connect-retries: 10

>     table: flyway_quarkus_history

Reviewer: If I'm not wrong, flyway will automatically create this table called `flyway_quarkus_history` - and as per requirements we need to provide a DDL scripts for PIM for the different databases to avoid auto-generation - so not sure if this really fits in our case. Also, customer will most likely want to execute the scripts by themselves.

I think we should discuss this

Author: Do you prefer to disable it by default?

    migrate-at-start: true
    baseline-on-migrate: true
    baseline-version: 1.0
    baseline-description: Quartz
  resteasy:
    path: /rest (1)
  datasource:
    db-kind: h2 (2)
    jdbc:
      url: jdbc:h2:mem:test;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
    username: sa
    password: sa
  hibernate-orm:
    database:
      generation: update
pim:
  auth-method: file (3)
```

1. Deploy the application on `/rest`
1. H2 inmemory datasource. Override it by one of your choice
1. Authentication method. Defaults to `file` but `jdbc` or `ldap` are also valid options

### Configuration overrides

Check [Quarkus Configuration Reference Guide](https://quarkus.io/guides/config-reference) for further details.

It is possible to override or extend the provided configuration. You can provide one or more additional configuration 
files that will allow you to customize the application. Several examples are provided in the [examples](./examples) 
folder.

As an example, if you want to replace the H2 default persistence configuration by 
[MariaDB](./examples/persistence/mariadb.yml) and the authentication mechanism to use 

> [LDAP](examples/authentication/ldap/ldap.yml).

Reviewer: Maybe to make it a little clearer, I would say something like:
```suggestion
As an example, if you want to replace the H2 default persistence configuration by 
[MariaDB](./examples/persistence/mariadb.yml) and the authentication mechanism to use 
[LDAP](examples/authentication/ldap/ldap.yml), see below.
```

> **Note:** As the MariaDB jdbc driver is not included in the classpath. It must be added.

Reviewer 2: Mention the JDBC driver extensions here: https://code.quarkus.redhat.com/?extension-search=jdbc%20driver

Author: Why not point here? https://quarkus.io/guides/datasource

Reviewer 2: Ah, that might be even better. If you link it somewhere already I might have missed that.

**Note:** These files will override or extend the already defined properties in the application.yaml file

#### Defining KIE Servers

The right way to configure the connection to one or more KIE Servers in order to perform the migrations, a list of 

> kieservers should exist in the configuration file.

Reviewer: ```suggestion
KIE servers should exist in the configuration file.
```

> kieservers should exist in the configuration file.

Reviewer: Maybe it would be useful to make a link to the configuration file you're talking about here. So it's clearer

...
``` 

## Build

It is a [Quarkus 2.x](https://quarkus.io/) application.

### JVM

```shell script
mvn clean package
```

### Native

**Note**: Native compilation is currently not supported.

```shell script
mvn clean package -Pnative -Dquarkus.native.container-build=true
```

## Run application

### JVM 

```bash
java -jar target/quarkus-app/quarkus-run.jar
```

### Native

```shell script
./target/process-migration-runner
```

You can provide your custom configuration file. Check [application.yaml](./src/main/resources/application.yaml) 
to see an example. The provided configuration will be added or override the one existing in application.yaml

### Container image

If you want to create a container image using your JVM or Native build use the `-Dquarkus.container-image.build=true` 
flag when packaging your application and pass in any other configuration property you want to override. See the 
[Container Image Guide](https://quarkus.io/guides/container-image#building).

```shell script 
$ mvn clean package -Dquarkus.container-image.build=true \
 -Dquarkus.container-image.push=true \
-Dquarkus.container-image.image=quay.io/kiegroup/process-migraiton-service:1.0
...

> [INFO] [io.quarkus.container.image.jib.deployment.JibProcessor] Created container image quay.io/kiegroup/process-migraiton-service:1.0 (sha256:dc8028963923081941529ef0ea515bd1e970b8fc96d5ad6a9346eb4ad61028f6)

Reviewer: ```suggestion
[INFO] [io.quarkus.container.image.jib.deployment.JibProcessor] Created container image quay.io/kiegroup/process-migration-service:1.0 (sha256:dc8028963923081941529ef0ea515bd1e970b8fc96d5ad6a9346eb4ad61028f6)
```

```

Or you can just use your favourite container builder tool and refer to any of the existing [Dockerfiles](./src/main/docker)

```shell script
$ podman build -t quay.io/kiegroup/process-migraiton-service:1.0 -f ./src/main/docker/Dockerfile.jvm .
```

## Configuration

Default configuration is as follows:

```yaml
quarkus:
  http:
    auth:
      basic: true
      policy:
        main-policy:
          roles-allowed: admin
      permission:
        main:
          paths: /*
          policy: main-policy
        public:
          paths: /q/health/*
          policy: permit
          methods: GET
  security:
    users:
      file:
        realm-name: pim_file
        enabled: true
        plain-text: true
        users: users.properties
        roles: roles.properties
    jdbc:

>       realm-name: Quarkus_jdbc

Reviewer: ```suggestion
      realm-name: pim_jdbc
```

## Run application

> It is a [Quarkus 1.x](https://quarkus.io/) application.

Reviewer 2: should be 2.x here, pom configuration is correct.

```

_Refer to the [Quarkus Datasource](https://quarkus.io/guides/datasource) configuration for further details_

#### Basic authentication and authorization

Authorization example available [here](./examples/authentication/basic-auth.yml). Shows how to define basic authorization.

**Note that this is a build time configuration** (See [Quarkus Authorization](https://quarkus.io/guides/security-authorization))

```{yaml}
  http:
    auth:
      basic: true
      policy:
        main-policy:
          roles-allowed: admin
      permission:
        main:
          paths: /*
          policy: main-policy
        public:
          paths: /q/health/*
          policy: permit
          methods: GET
```

1. Authentication type `BASIC`
1. Every resource under root path requires the `admin` role
1. Health checks are not secured

By default, the user properties IdentityProvider is provided in plain text:

**Note that this is a build time configuration** (See [Quarkus Security](https://quarkus.io/guides/security#identity-providers))

```{yaml}
  security:
    users:
      file:
        enabled: true
        plain-text: true
        users: users.properties
        roles: roles.properties
```

> ## Using non-provided JDBC drivers

Reviewer: Given the fact that we may release this with only H2 only (or even only sources) I think it would be worth giving a few commands or examples about how to add these dependencies or execute this command. 
- Is it needed to re-build the app with the driver in the classpath?
- or is it just enough to pass some quarkus parameters at the start-up in the same way we used to do this with thorntail before? like `java -Dthorntail.classpath=./mariadb-java-client-2.4.2.jar -jar target/process-migration-thorntail.jar -s./mariadb-config.yml`

> The H2 JDBC driver is included by default. However, users will want to use different JDBC drivers to connect to external databases. 

Reviewer: ```suggestion
The H2 JDBC driver is included by default. However, users will be able to use different JDBC drivers to connect to external databases. 
```

> For that purpose you will have add a dependency to your preferred database. Check all the supported Database drivers in

Reviewer 2: ```suggestion
For that purpose you will have add a Quarkus extension to your preferred database. Check all the supported JDBC drivers extensions in
```

> For that purpose you will have add a dependency to your preferred database. Check all the supported Database drivers in

Reviewer: ```suggestion
For that purpose you will have to add a dependency for your preferred database. Check all the supported Database drivers in
```

    password: secret
```

#### MariaDB Datasource

> See [Using non-provided JDBC drivers](#using-non-provided-jdbc-drivers) for details on how to include additional JDBC drivers to the runtime.

Reviewer: This needs to be updated
```suggestion
See [Using other JDBC extensions](#using-other-jdbc-extensions) for details on how to include additional JDBC drivers to the runtime.
```

```yaml
quarkus:
  datasource:
    db-kind: mariadb
    jdbc:
      url: jdbc:mariadb://localhost:3306/pimdb
    username: pim
    password: pim123
```

_Refer to the [Quarkus Datasource](https://quarkus.io/guides/datasource) configuration for further details_

#### Basic authentication and authorization

Authorization example available [here](./examples/authentication/basic-auth.yml). Shows how to define basic authorization.

**Note that this is a build time configuration** (See [Quarkus Authorization](https://quarkus.io/guides/security-authorization))

> ```{yaml}

Reviewer: ```suggestion
```yaml
```

  http:
    auth:
      basic: true
      policy:
        main-policy:
          roles-allowed: admin
      permission:
        main:
          paths: /*
          policy: main-policy
        public:
          paths: /q/health/*
          policy: permit
          methods: GET
```

1. Authentication type `BASIC`
1. Every resource under root path requires the `admin` role
1. Health checks are not secured

> By default, the user properties IdentityProvider is provided in plain text:

Reviewer: rather than plain text, it's actually provided through file based in this example

Author: That means that the properties files are not encoded by default.

**Note that this is a build time configuration** (See [Quarkus Security](https://quarkus.io/guides/security#identity-providers))

> ```{yaml}

Reviewer: ```suggestion
```yaml
```

Response:

```http

> Status: 201 CREATED

Reviewer: Are we changing the status code in the response with this release? or it was just because the readme file was not updated?

Author: I decided to make take the opportunity to make status code more accurate. I can leave this for a next iteration

Response:

```http

> Status: 201 CREATED

Reviewer: same as above
